[AMF] Support for timers t3560, t3550 and t3570

Kaleem-Wavelabs opened the pull request:
<!--
    Tag your PR title with the components that it touches.
    E.g. "[lte][agw] Changeset" or "[orc8r][docker] ..."
-->

## Summary

Fixes:
Implemented the timer wrapper APIs to maintain the thread arguments to be consistent in callbacks
Added the missing code functionality to support T3560 timer, t3550 and t3570


## Test Plan

Test cases:
1) Run with UERANSIM to miss the response for authentication request
2) Run with UERANSIM to send the authentication success and failure messages after initial failures
3) Run with UERANSIM to miss the response for registration request
4) Run with UERANSIM to send the registration success and failure messages after initial failures

Signed-off-by: Kaleemullah Mohammed <kaleemullah.mohammed@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->
![authenticationrequest](https://user-images.githubusercontent.com/83354949/123457446-8c2cf500-d601-11eb-867a-0db43dd36444.jpg)
![securitymode](https://user-images.githubusercontent.com/83354949/123457451-8df6b880-d601-11eb-81e1-0e99327416f4.jpg)
![successfulcase](https://user-images.githubusercontent.com/83354949/123457638-bed6ed80-d601-11eb-909f-74d0651cadb1.jpg)

Commit:
Revert "Fixes:"

This reverts commit f171554e6be5a44ebf560d938c2d96ec855f25e1.

Commit:
Support for timers t3560, t3550 and t3570

Signed-off-by: vagrant <vagrant@magma-dev>
Signed-off-by: Kaleemullah Mohammed <kaleemullah.mohammed@wavelabs.ai>

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -596,10 +596,9 @@ int amf_send_registration_accept(amf_context_t* amf_context) {
-          (void*) registration_proc->ue_id);
+      registration_proc->T3550.id = amf_app_start_timer(
+          REGISTRATION_ACCEPT_TIMER_EXPIRY_MSECS, TIMER_REPEAT_ONCE,
+          registration_accept_t3550_handler, registration_proc->ue_id);
```
do we need 2 log statements? timer start, timer started. log statements cause I/O overhead on the network function. let us takeout redundant log statement.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -652,23 +656,21 @@ static int registration_accept_t3550_handler(
       OAILOG_ERROR(
           LOG_AMF_APP,
           "Timer: timer has expired Sending Registration request again\n");
```
do you want to indicate timer ID or timer T3XXX in the context of UE XXXX expired? is this for registration Request? can you double check? is AMF sending registration request in this case?

Kaleem-Wavelabs replied:
Updated the log statement accordingly

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -652,23 +656,21 @@ static int registration_accept_t3550_handler(
-      // amf_registration_request(registration_proc); /* TO DO for negative
+      amf_send_registration_accept(amf_ctx);
       // scenario */
+      OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNok);
```
what gets logged here? is this necessary?

sureshgorijavolu replied:
what is this TODO? can we fix it?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -652,23 +656,21 @@ static int registration_accept_t3550_handler(
-
       OAILOG_ERROR(
           LOG_AMF_APP,
           "Timer: Maximum retires done hence Abort the registration "
```
can you indicate Timer type and Number of retries finished by now?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -652,23 +656,21 @@ static int registration_accept_t3550_handler(
           "Timer: Maximum retires done hence Abort the registration "
           "procedure\n");
-      return -1;
+      amf_remove_ue_context(ue_amf_context);
```
what does the spec 24.501 suggest? is it enough if we remove UE context? or do we need to send any downlink NAS message in this case?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/Registration.cpp:
```diff
@@ -762,7 +764,7 @@ int amf_proc_registration_complete(
-      stop_timer(&amf_app_task_zmq_ctx, registration_proc->T3550.id);
+      amf_app_stop_timer(registration_proc->T3550.id);
       OAILOG_INFO(
           LOG_AMF_APP, "Timer: after stop registration timer with id = %d\n",
```
here id is mentioned good. in addition to ID, can you indicate Timer T3xxx in the log statement?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_Security_Mode.cpp:
```diff
@@ -68,7 +68,7 @@ int amf_handle_security_complete_response(
 
-    stop_timer(&amf_app_task_zmq_ctx, smc_proc->T3560.id);
+    amf_app_stop_timer(smc_proc->T3560.id);
     OAILOG_DEBUG(LOG_AMF_APP, "Timer: After stopping SMC MODE timer \n");
```
let us not indicate with informal naming for the timer such as "SMC MODE timer" or something like. let us use an uniform naming T3xxx timer.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_app_timer_management.h:
```diff
@@ -0,0 +1,69 @@
+
+void amf_app_stop_timer(int timer_id);
+
+/*void amf_app_resume_timer(
```
if you no longer need this, you can take out this dead code.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -192,13 +192,18 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
   OAILOG_INFO(LOG_AMF_APP, "Timer: In security_mode_t3560_handler - T3560\n");
```
log statements are not using an unifrom naming convention. it is better to fix it.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -192,13 +192,18 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+
+  if (!amf_app_get_timer_arg(timer_id, &ue_id)) {
+    OAILOG_WARNING(
+        LOG_AMF_APP, "Invalid Timer Id expiration, Timer Id: %u\n", timer_id);
```
simplify the language.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -192,13 +192,18 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
   ue_m5gmm_context_s* ue_context = amf_ue_context_exists_amf_ue_ngap_id(ue_id);
 
   if (ue_context == NULL) {
     OAILOG_INFO(LOG_AMF_APP, "AMF_TEST: ue_context is NULL\n");
```
include ue_id in the log statement. it will be handy. is this INFO level log?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -208,8 +213,7 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
   OAILOG_INFO(LOG_AMF_APP, "Timer: got amf ctx and calling common procedure\n");
```
simplify log statement

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
         LOG_AMF_APP, "T3560 timer   timer id %d ue id %d\n", smc_proc->T3560.id,
```
this statement is providing sufficient info for troubleshooting. other statements above can also follow this approach.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+
+    if (smc_proc->retransmission_count < SECURITY_COUNTER_MAX) {
+      /*
+       * Send identity request message to the UE
```
it is not sending, it is Retransmission. let us fix the statement to improve readability to the coders / maintainers. it is just retransmission. So no need to repopulate the message fields using RAND | AUTN etc.,

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+       */
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Timer: timer has expired Sending Security Command Mode request "
```
mention timer T3xxx and its ID in the statement.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+          LOG_AMF_APP,
+          "Timer: timer has expired Sending Security Command Mode request "
+          "again\n");
+      amf_security_request(smc_proc);
```
a better name can be given to the function. is it Security Mode Command? instead of security_request can this be renamed to smc or security_mode_cmd? it is not request, it is a command.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+       */
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Timer: Maximum retires done hence Abort the smc procedure\n");
```
let us indicate "Max Number of retries  x for Timer T3xxx expired."

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp:
```diff
@@ -221,36 +225,37 @@ static int security_mode_t3560_handler(zloop_t* loop, int timer_id, void* arg) {
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Timer: Maximum retires done hence Abort the smc procedure\n");
+      amf_remove_ue_context(ue_context);
```
is it enough if UE context is removed? do we need to send any Downlink message to UE. refer to procedure in 24.501

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/nas_proc.cpp:
```diff
@@ -263,9 +263,9 @@ static int amf_identification_request(nas_amf_ident_proc_t* const proc) {
     // TODO
```
can we take out the TODO comment from this file?

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/nas_proc.cpp:
```diff
@@ -303,38 +307,39 @@ static int identification_t3570_handler(
+          IDENTIFICATION_COUNTER_MAX);
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Timer: timer has expired Sending Identification request again\n");
```
statement should say "retransmitting"

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/nas_proc.cpp:
```diff
@@ -303,38 +307,39 @@ static int identification_t3570_handler(
+       */
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Timer: Maximum retires done hence Abort the identification "
```
improve readability as indicated in the examples above.

sureshgorijavolu commented on lte/gateway/c/oai/tasks/amf/nas_proc.cpp:
```diff
@@ -303,38 +307,39 @@ static int identification_t3570_handler(
   }
-#endif /*  TIMER_CHANGES_REVIEW */
-  return 0;
+  OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNerror);
```
it is possible to drive these retransmissions through state machine. it is important to drive it all through state machine.

Kaleem-Wavelabs commented:
Shared the mme binary to TeraVM for verification of the change.
Addressed the review comments shared.
Will be amending the changes once the test results are shared
